tool to generate constraint approximation databases

v4hn opened the pull request:
This feature was on my todo list for quite some time now.
MoveIt provides [some C++ demo code](https://github.com/ros-planning/moveit/blob/melodic-devel/moveit_planners/ompl/ompl_interface/src/demo_construct_state_database.cpp) to generate constraint approximations.
These can be used for [constraint path planning](https://ros-planning.github.io/moveit_tutorials/doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.html).
The demo is all hard-coded though and users had to copy&paste the code for their own use case.

This pull-request provides a way to [describe constraints on the parameter server](https://github.com/ros-planning/moveit/commit/783cf74a76350692877e6327df247ea4b18f6861) and applies the feature to configure [a generic tool](https://github.com/ros-planning/moveit/commit/3fc0789c6c6da9573cc5134ab2b11e39c7bc5202) to generate the databases.

I'm open to change & move parts of this code if someone has a better proposal for where things should go.

Please look at the individual commits for more details.

v4hn commented:
I plan to extend [the constraint manifold tutorial](https://ros-planning.github.io/moveit_tutorials/doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.html) once I got an approval on the current design.

I'm not sure where to document the structure for the `constructConstraints` function best, or whether it needs to be documented explicitly when the manifold-tutorial gives examples already.
Maybe it makes sense to add a link in the function's doxygen documentation?

Thanks for the feedback, I will address it later this week.

v4hn commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -269,4 +272,230 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
+    }
+    else if (it->first == "tolerances")
+    {
+      if (it->second.getType() != XmlRpc::XmlRpcValue::TypeArray || it->second.size() != 2)
```
Good idea, but I wanted to leave the possibility to possibly check for multiple different structures in the future.

I factorized a `isArray` and an `isStruct` method, but made the logging optional.

v4hn commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -269,4 +272,230 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
+      geometry_msgs::Pose& region_pose = constraint.constraint_region.primitive_poses.back();
+      shape_msgs::SolidPrimitive& region_primitive = constraint.constraint_region.primitives.back();
+
+      std::function<double(XmlRpc::XmlRpcValue&)> center = [](XmlRpc::XmlRpcValue& it) {
```
I prefer to keep this as-is for better readability.

rhaschke replied:
For me, your version wasn't readable. It took me a while to figure out, that x,y,z components are all "parsed" twice, once with `center()` and once with `extend()`. Splitting this into two functions disguises the fact that center and extend are both parsed from the very same x,y,z elements, which are ranges actually.

v4hn replied:
I guess it depends on whether you look at it as "parse ros-parameters" or as "fill in a full constraint message". I don't care too much, so I changed it as suggested.

v4hn commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -269,4 +272,230 @@ moveit_msgs::Constraints constructGoalConstraints(const std::string& link_name,
+
+bool constructConstraint(XmlRpc::XmlRpcValue& params, moveit_msgs::VisibilityConstraint& constraint)
+{
+  ROS_FATAL_NAMED(LOGNAME, "TODO: VisibilityConstraint parsing from parameters is not implemented");
```
To be honest I believe the current implementation of this constraint is far from elegant and I have not seen a single case where this was used in practice..
So I just did not bother before.

I added (and verified) code for this in another commit now.

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+  }
+  if (nh.param<bool>("use_current_scene", true))
+  {
+    ROS_INFO("Requesting current planning scene to generate database");
```
NAMED

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+    ROS_INFO("Requesting current planning scene to generate database");
+    if (!psm.requestPlanningSceneState())
+    {
+      ROS_FATAL("Abort. The current scene could not be retrieved.");
```
NAMED

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+  opt.max_explicit_points = nh.param<int>("max_explicit_points", 200);
+  opt.state_space_parameterization = nh.param<std::string>("state_space_parameterization", "JointModel");
+
+  ROS_INFO_STREAM("Generating Joint Space Constraint Approximation Database for constraint:\n" << constraints);
```
NAMED

mlautman replied:
Why in the middle of reading parameters? Why not keep all parameters together?

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+
+  ROS_INFO_STREAM("Generating Joint Space Constraint Approximation Database for constraint:\n" << constraints);
+
+  const std::string planning_group = nh.param<std::string>("planning_group", "<unspecified planning group>");
```
Is this actually a good default? Maybe return with an error msg would be better here

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+      ompl_interface.getConstraintsLibrary().addConstraintApproximation(constraints, planning_group, scene, opt);
+  if (!result.approx)
+  {
+    ROS_FATAL("Failed to generate approximation.");
```
NAMED

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+  if (!nh.getParam(CONSTRAINT_PARAMETER, constraint_description) ||
+      !kinematic_constraints::constructConstraints(constraint_description, constraints))
+  {
+    ROS_FATAL_STREAM("Could not find valid constraint description in parameter '"
```
NAMED

mlautman commented on moveit_planners/ompl/ompl_interface/src/generate_state_database.cpp:
```diff
@@ -0,0 +1,140 @@
+  // number of states in joint space approximation
+  opt.samples = nh.param<int>("state_cnt", 10000);
+
+  // useful default parameterization
```
Seems like making a class and putting all the parameter reading into a separate method would be cleaner

v4hn commented:
@rhaschke @mlautman I addressed all feedback. Please have another look.

@mlautman Personally, I prefer in-place `nh.param` calls over parameter structs in such small programs, but I switched to a struct here.

rhaschke commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -273,3 +274,298 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
+      shape_msgs::SolidPrimitive& region_primitive = constraint.constraint_region.primitives.back();
+
+      std::function<double(XmlRpc::XmlRpcValue&)> center = [](XmlRpc::XmlRpcValue& it) {
+        return parseDouble(it[0]) + (parseDouble(it[1]) - parseDouble(it[0])) / 2;
```
This is hard to read as well. The center of the interval is just `(parseDouble(it[0]) + parseDouble(it[1])) / 2;`

rhaschke commented on moveit_core/kinematic_constraints/src/utils.cpp:
```diff
@@ -273,3 +274,298 @@ moveit_msgs::Constraints kinematic_constraints::constructGoalConstraints(const s
   return goal;
 }
+
+static double parseDouble(XmlRpc::XmlRpcValue& v)
```
Please provide some const specifiers to the arguments.

rhaschke replied:
Please neglect my comment. Obviously XmlRpc doesn't correctly declare constness of cast operations.

v4hn replied:
Yes, the `XmlRpc` API is rather badly designed. This shows in a number of places in the request.
`int` for array indices, no iterators for arrays, missing `const` specifies in API functions, ...

rhaschke commented:
@v4hn Do you want to squash the cleanup commits into the first 4 commits?

v4hn commented:
I will clean up the commit history and merge.

v4hn commented:
merged I'll try to get around to add the tutorial soon.
